### PR TITLE
Implement Serial DMA abstraction.

### PR DESCRIPTION
**Pull request by ZhiyaoMa98**
The implementation follows existing DMA abstraction for I2C interface.

The implementation has been tested on STM32F407-Discovery board with the code in `examples/serial-dma.rs`.

### CODE REVIEW

File: src/serial/dma.rs [hunk 1]
```diff
@@ -0,0 +1,711 @@
...
+    /// # Warning
+    /// `callback` may be called before function returns value. It happens on errors in preparation stages.
+    unsafe fn read_dma(
+        &mut self,
+        buf: &mut [u8],
+        callback: Option<SerialCompleteCallback>,
+    ) -> nb::Result<(), super::Error>;
+}
+
+pub trait SerialWriteReadDMA {
```
Comment (burrbull) [inline src/serial/dma.rs:64]: What use-case of WriteRead in with Serial?

Comment (Acciente717) [reply src/serial/dma.rs:64]: To be honest, I don't have a current use case for this. I was basically following the existing implementation of I2C and made it to similarly work for serial.

However, I do think there can be use case for such communication. For example, when communicating with another chip connected over serial, the WriteRead can be used to support the paradigm of send query and read response from the external chip.

Comment (burrbull) [reply src/serial/dma.rs:64]: I2C is synchronous interface. UART is asynchronous.  This paradigm will not work properly without synchronization.

Please, remove this part from PR.

Comment (ZhiyaoMa98) [reply src/serial/dma.rs]: Thanks. `SerialWriteReadDMA` is removed.

**Review by burrbull — APPROVED** [review]

**Comment by burrbull** [comment]
Sorry. One more. Do not use `merge master` in PR, please.
`rebase` it instead.

**Comment by Acciente717** [comment]
No problem. Done.

**Comment by Acciente717** [comment]
Ah sorry seems like I brought back the `SerialWriteReadDMA` trait. Let me remove it again.

**Review by burrbull — APPROVED** [review]

**Pull request merged**